### DISCUSSION TITLE
Should `@babel` presets need to be normal dependencies in order for `assets:precompile` to work?

### DISCUSSION THREAD

**Discussion by jennysharps** [shakacode/shakapacker#244] [General]
We are working on moving to `shakapacker` from [`webpacker@6.0.0.rc.6`](https://github.com/rails/webpacker/tree/v6.0.0.rc.6) and noticed that the [installation instructions](https://github.com/shakacode/shakapacker/tree/master#installation) advise us to install the `@babel/preset-env` package--among others--as normal dependency instead of as a dev dependency.

In addition to `@babel/preset-env`, we also use `@babel/preset-react` &  `@babel/preset-typescript`. Usually, all of these presets are installed as `devDependencies` as per https://babeljs.io/docs/en/babel-preset-env#install, for example.  When we moved them to `devDependencies`, the `assets:precompile` step began to fail with the following in our CI pipeline:

```
[0mI, [2023-02-01T17:26:51.505750 <redacted>]  INFO -- : Compiling...
E, [2023-02-01T17:26:54.871106 <redacted>] ERROR -- : 
COMPILATION FAILED:
EXIT STATUS: pid 262 exit 1
OUTPUTS:
node:internal/process/promises:279
            triggerUncaughtException(err, true /* fromPromise */);
            ^

Error [ERR_MODULE_NOT_FOUND]: Cannot find package '@babel/preset-react' imported from /app/babel-virtual-resolve-base.js
    at new NodeError (/app/node_modules/@babel/core/lib/vendor/import-meta-resolve.js:2240:5)
    at packageResolve (/app/node_modules/@babel/core/lib/vendor/import-meta-resolve.js:2776:9)
    at moduleResolve (/app/node_modules/@babel/core/lib/vendor/import-meta-resolve.js:2804:18)
    at defaultResolve (/app/node_modules/@babel/core/lib/vendor/import-meta-resolve.js:2835:13)
    at /app/node_modules/@babel/core/lib/vendor/import-meta-resolve.js:2855:14
    at Generator.next (<anonymous>)
    at asyncGeneratorStep (/app/node_modules/@babel/core/lib/vendor/import-meta-resolve.js:45:103)
    at _next (/app/node_modules/@babel/core/lib/vendor/import-meta-resolve.js:46:194)
    at /app/node_modules/@babel/core/lib/vendor/import-meta-resolve.js:46:364
    at new Promise (<anonymous>) {
  code: 'ERR_MODULE_NOT_FOUND'
}

The command '/bin/sh -c CI=*****
Service 'rails-app' failed to build : Build failed
```

It works fine if we promote them to `dependencies`. Is this expected?  Can anybody shed some light on this behavior?

**Comment by jennysharps** [comment]
I believe this was happening because we were running `RAILS_ENV=production bundle exec rake assets:precompile`.  After moving the presets back to `devDependencies` and not setting `RAILS_ENV` at all when precompiling assets, the precompile step succeeds.

What is the intended way to precompile assets for production?  Even when `RAILS_ENV` is set to production, I would've though a full install would still be required since dev dependencies are needed to build the bundle. An example is compiling typescript -- the `@types` packages are dev dependencies, but we need them in order to do type-checking when building the production bundle; we just don't need them to end up in the production bundle.

**Comment by jennysharps** [comment]
I believe the problem was multiple `yarn install`s happening during precompilation (still not sure why they installed production packages only 🤷‍♀️), and this seems to have fixed it https://forum.shakacode.com/t/how-to-skip-automatically-running-yarn-install-when-you-precompile/2417

**Comment by justin808** [comment]
@jennysharps I think the latest versions don't do the yarn installs at all. It's important to have your build process call `yarn install` to install the npm dependencies, just like the gems get installed.

Regarding production packages, many build environments run with the flag to skip the devDependencies. For example, that is the default behavior of Heroku's buildpacks.
